File: core/math/src/main/java/com/astroforge/dysonprotocol/core/math/BigNumber.kt
package com.astroforge.dysonprotocol.core.math

import kotlin.math.floor
import kotlin.math.log10
import kotlin.math.pow
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

/**
 * Scientific-notation numeric type for idle-game magnitudes far beyond Double (~1e308).
 *
 * Representation: `value = mantissa × 10^exponent` where mantissa is 0 or in ±[1, 10).
 * Operations allocate a single result instance; hot paths should reuse [BigNumberScratch]
 * when a loop would otherwise churn wrappers.
 */
data class BigNumber(
    val mantissa: Double,
    val exponent: Long,
) : Comparable<BigNumber> {

    init {
        require(mantissa.isFinite()) { "mantissa must be finite, was $mantissa" }
    }

    val isZero: Boolean get() = mantissa == 0.0
    val isPositive: Boolean get() = mantissa > 0.0
    val isNegative: Boolean get() = mantissa < 0.0
    val sign: Int get() = when {
        mantissa > 0.0 -> 1
        mantissa < 0.0 -> -1
        else -> 0
    }

    operator fun plus(other: BigNumber): BigNumber = add(other)

    operator fun minus(other: BigNumber): BigNumber = add(other.unaryMinus())

    operator fun times(other: BigNumber): BigNumber {
        if (isZero || other.isZero) return ZERO
        return fromScientific(mantissa * other.mantissa, exponent + other.exponent)
    }

    operator fun times(scalar: Double): BigNumber {
        if (isZero || scalar == 0.0) return ZERO
        require(scalar.isFinite()) { "scalar must be finite" }
        return fromScientific(mantissa * scalar, exponent)
    }

    operator fun times(scalar: Int): BigNumber = times(scalar.toDouble())

    operator fun times(scalar: Long): BigNumber = times(scalar.toDouble())

    operator fun div(other: BigNumber): BigNumber {
        if (isZero) return ZERO
        if (other.isZero) return ZERO
        return fromScientific(mantissa / other.mantissa, exponent - other.exponent)
    }

    operator fun div(scalar: Double): BigNumber {
        if (isZero) return ZERO
        if (scalar == 0.0) return ZERO
        require(scalar.isFinite()) { "scalar must be finite" }
        return fromScientific(mantissa / scalar, exponent)
    }

    operator fun unaryMinus(): BigNumber = if (isZero) ZERO else copy(mantissa = -mantissa)

    operator fun inc(): BigNumber = this + ONE

    fun abs(): BigNumber = if (mantissa < 0.0) copy(mantissa = -mantissa) else this

    fun floor(): BigNumber {
        if (isZero) return ZERO
        if (exponent < 0L) return if (isPositive) ZERO else -ONE
        if (exponent >= 15L) return this
        val value = mantissa * 10.0.pow(exponent.toDouble())
        return fromDouble(floor(value))
    }

    /**
     * Integer power via exponentiation by squaring. Negative exponents invert the value.
     */
    fun pow(power: Int): BigNumber {
        if (isZero) return if (power > 0) ZERO else if (power == 0) ONE else ZERO
        if (power == 0) return ONE
        if (power == 1) return this
        if (power < 0) return ONE / pow(-power)
        var result = ONE
        var base = this
        var exp = power
        while (exp > 0) {
            if (exp and 1 == 1) result *= base
            exp = exp ushr 1
            if (exp > 0) base *= base
        }
        return result
    }

    /**
     * Real power using logarithms: a^p = 10^(p · log10(a)).
     * Negative bases are rejected; zero to a positive power is zero.
     */
    fun pow(power: Double): BigNumber {
        require(power.isFinite()) { "power must be finite" }
        if (power == 0.0) return ONE
        if (isZero) return if (power > 0.0) ZERO else ZERO
        require(!isNegative) { "real power is undefined for negative BigNumber" }
        if (power == 1.0) return this
        val log = log10()
        val resultLog = log * power
        if (!resultLog.isFinite()) {
            return if (resultLog > 0.0) fromScientific(1.0, Long.MAX_VALUE / 4) else ZERO
        }
        val expPart = floor(resultLog)
        var exp = expPart.toLong()
        var man = 10.0.pow(resultLog - expPart)
        if (!man.isFinite() || man == 0.0) {
            return if (resultLog > 0.0) fromScientific(1.0, Long.MAX_VALUE / 4) else ZERO
        }
        while (man >= 10.0) {
            man /= 10.0
            exp += 1
        }
        while (man > 0.0 && man < 1.0) {
            man *= 10.0
            exp -= 1
        }
        return BigNumber(man, exp)
    }

    fun log10(): Double {
        if (isZero || isNegative) return Double.NEGATIVE_INFINITY
        return log10(mantissa) + exponent.toDouble()
    }

    fun ln(): Double = log10() * LN_10

    fun toDouble(): Double {
        if (isZero) return 0.0
        if (exponent > 308L) return if (isPositive) Double.POSITIVE_INFINITY else Double.NEGATIVE_INFINITY
        if (exponent < -308L) return 0.0
        return mantissa * 10.0.pow(exponent.toDouble())
    }

    fun toLongClamped(): Long {
        val floored = floor()
        if (floored.isZero) return 0L
        if (floored.exponent > 18L) return if (isPositive) Long.MAX_VALUE else Long.MIN_VALUE
        val v = floored.toDouble()
        return if (v > Long.MAX_VALUE) Long.MAX_VALUE else if (v < Long.MIN_VALUE) Long.MIN_VALUE else v.toLong()
    }

    fun formatScientific(fractionDigits: Int = 3): String = BigNumberFormat.scientific(this, fractionDigits)

    fun formatEngineering(fractionDigits: Int = 2): String = BigNumberFormat.engineering(this, fractionDigits)

    fun formatSuffix(fractionDigits: Int = 2): String = BigNumberFormat.suffix(this, fractionDigits)

    override fun compareTo(other: BigNumber): Int {
        if (isZero && other.isZero) return 0
        val signCmp = sign.compareTo(other.sign)
        if (signCmp != 0) return signCmp
        if (isZero) return 0
        val expCmp = exponent.compareTo(other.exponent)
        if (expCmp != 0) return if (isPositive) expCmp else -expCmp
        return mantissa.compareTo(other.mantissa)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is BigNumber) return false
        if (isZero && other.isZero) return true
        return exponent == other.exponent && abs(mantissa - other.mantissa) < 1e-12
    }

    override fun hashCode(): Int {
        if (isZero) return 0
        return 31 * mantissa.hashCode() + exponent.hashCode()
    }

    override fun toString(): String = formatScientific()

    companion object {
        val ZERO = BigNumber(0.0, 0L)
        val ONE = BigNumber(1.0, 0L)
        val TEN = BigNumber(1.0, 1L)
        val HUNDRED = BigNumber(1.0, 2L)
        val THOUSAND = BigNumber(1.0, 3L)
        val MILLION = BigNumber(1.0, 6L)
        val E12 = BigNumber(1.0, 12L)

        private const val LN_10 = 2.302585092994046
        private const val ALIGN_LIMIT = 18

        fun fromDouble(value: Double): BigNumber {
            if (value == 0.0 || !value.isFinite()) {
                require(value.isFinite() || value == 0.0) { "non-finite double: $value" }
                return ZERO
            }
            return fromScientific(value, 0L)
        }

        fun fromLong(value: Long): BigNumber {
            if (value == 0L) return ZERO
            return fromDouble(value.toDouble())
        }

        fun fromInt(value: Int): BigNumber = fromLong(value.toLong())

        /**
         * Builds a normalized value from a raw mantissa/exponent pair.
         */
        fun fromScientific(rawMantissa: Double, rawExponent: Long): BigNumber {
            if (rawMantissa == 0.0 || !rawMantissa.isFinite()) return ZERO
            var man = rawMantissa
            var exp = rawExponent
            val negative = man < 0.0
            man = abs(man)
            val log = log10(man)
            if (log.isFinite()) {
                val shift = floor(log).toLong()
                man /= 10.0.pow(shift.toDouble())
                exp += shift
            } else {
                while (man >= 10.0) {
                    man /= 10.0
                    exp += 1
                    if (exp == Long.MAX_VALUE) break
                }
                while (man > 0.0 && man < 1.0) {
                    man *= 10.0
                    exp -= 1
                    if (exp == Long.MIN_VALUE) break
                }
            }
            if (man >= 10.0) {
                man /= 10.0
                exp += 1
            }
            if (man > 0.0 && man < 1.0) {
                man *= 10.0
                exp -= 1
            }
            if (negative) man = -man
            return BigNumber(man, exp)
        }

        operator fun invoke(value: Double): BigNumber = fromDouble(value)
        operator fun invoke(value: Int): BigNumber = fromInt(value)
        operator fun invoke(value: Long): BigNumber = fromLong(value)
    }

    private fun add(other: BigNumber): BigNumber {
        if (isZero) return other
        if (other.isZero) return this
        val expDelta = exponent - other.exponent
        if (expDelta >= ALIGN_LIMIT) return this
        if (expDelta <= -ALIGN_LIMIT) return other
        val (hi, lo) = if (exponent >= other.exponent) this to other else other to this
        val scale = 10.0.pow((lo.exponent - hi.exponent).toDouble())
        val summed = hi.mantissa + lo.mantissa * scale
        return fromScientific(summed, hi.exponent)
    }
}

operator fun Double.times(other: BigNumber): BigNumber = other * this
operator fun Int.times(other: BigNumber): BigNumber = other * this
operator fun Long.times(other: BigNumber): BigNumber = other * this

/**
 * Mutable accumulator for high-frequency loops that must not allocate per tick.
 * Not thread-safe. Convert to [BigNumber] only when publishing state.
 */
class BigNumberScratch {
    var mantissa: Double = 0.0
    var exponent: Long = 0L

    fun set(value: BigNumber) {
        mantissa = value.mantissa
        exponent = value.exponent
    }

    fun setZero() {
        mantissa = 0.0
        exponent = 0L
    }

    fun toBigNumber(): BigNumber = BigNumber.fromScientific(mantissa, exponent)

    fun add(other: BigNumber) {
        if (mantissa == 0.0) {
            set(other)
            return
        }
        if (other.isZero) return
        val delta = exponent - other.exponent
        if (delta >= 18) return
        if (delta <= -18) {
            set(other)
            return
        }
        if (delta >= 0) {
            mantissa += other.mantissa * 10.0.pow(-delta.toDouble())
        } else {
            mantissa = mantissa * 10.0.pow(delta.toDouble()) + other.mantissa
            exponent = other.exponent
        }
        normalizeInPlace()
    }

    fun mulScalar(scalar: Double) {
        mantissa *= scalar
        normalizeInPlace()
    }

    private fun normalizeInPlace() {
        if (mantissa == 0.0 || !mantissa.isFinite()) {
            mantissa = 0.0
            exponent = 0L
            return
        }
        val negative = mantissa < 0.0
        var man = abs(mantissa)
        val log = log10(man)
        if (log.isFinite()) {
            val shift = floor(log).toLong()
            man /= 10.0.pow(shift.toDouble())
            exponent += shift
        }
        if (negative) man = -man
        mantissa = man
    }
}

File: src/core/math/big-number.ts
/**
 * Mantissa-exponent scientific number for idle-game magnitudes far beyond 1e308.
 * Immutable. Hot-path ops allocate one object; they never recurse or loop by
 * magnitude. Mantissa is normalized to 0 or [1, 10) with sign.
 */
const LOG10 = Math.LN10;
const INV_LOG10 = 1 / LOG10;
const EXP_ABS_CAP = 1e12;

function clampExp(e: number): number {
  if (e > EXP_ABS_CAP) return EXP_ABS_CAP;
  if (e < -EXP_ABS_CAP) return -EXP_ABS_CAP;
  return e;
}

export class BigNumber {
  readonly m: number;
  readonly e: number;

  constructor(mantissa: number, exponent = 0) {
    if (!Number.isFinite(mantissa) || mantissa === 0) {
      if (mantissa === Infinity) {
        this.m = 1;
        this.e = EXP_ABS_CAP;
        return;
      }
      if (mantissa === -Infinity) {
        this.m = -1;
        this.e = EXP_ABS_CAP;
        return;
      }
      this.m = 0;
      this.e = 0;
      return;
    }
    if (!Number.isFinite(exponent)) {
      this.m = mantissa < 0 ? -1 : 1;
      this.e = EXP_ABS_CAP;
      return;
    }

    const sign = mantissa < 0 ? -1 : 1;
    let mag = Math.abs(mantissa);
    let exp = exponent;

    // log10 path — constant time, no loops over the exponent.
    const log = Math.log(mag) * INV_LOG10;
    const shift = Math.floor(log);
    mag = mag * Math.pow(10, -shift);
    exp = clampExp(exp + shift);

    if (mag >= 10) {
      mag /= 10;
      exp = clampExp(exp + 1);
    } else if (mag < 1) {
      mag *= 10;
      exp = clampExp(exp - 1);
    }

    this.m = sign * mag;
    this.e = exp;
  }

  static readonly ZERO = new BigNumber(0, 0);
  static readonly ONE = new BigNumber(1, 0);
  static readonly TEN = new BigNumber(10, 0);

  static from(n: number): BigNumber {
    if (n === 0) return BigNumber.ZERO;
    if (n === 1) return BigNumber.ONE;
    return new BigNumber(n, 0);
  }

  static fromParts(m: number, e: number): BigNumber {
    return new BigNumber(m, e);
  }

  static fromJSON(raw: unknown): BigNumber {
    if (raw instanceof BigNumber) return raw;
    if (typeof raw === "number") return BigNumber.from(raw);
    if (raw && typeof raw === "object" && "m" in raw && "e" in raw) {
      const rec = raw as { m: unknown; e: unknown };
      return new BigNumber(Number(rec.m) || 0, Number(rec.e) || 0);
    }
    if (typeof raw === "string") return BigNumber.parse(raw);
    return BigNumber.ZERO;
  }

  static parse(text: string): BigNumber {
    const t = text.trim();
    const sci = t.match(/^([+-]?\d+(?:\.\d+)?)e([+-]?\d+)$/i);
    if (sci) return new BigNumber(Number(sci[1]), Number(sci[2]));
    const n = Number(t);
    return Number.isFinite(n) ? BigNumber.from(n) : BigNumber.ZERO;
  }

  static min(a: BigNumber, b: BigNumber): BigNumber {
    return a.lt(b) ? a : b;
  }

  static max(a: BigNumber, b: BigNumber): BigNumber {
    return a.gt(b) ? a : b;
  }

  isZero(): boolean {
    return this.m === 0;
  }

  isFinite(): boolean {
    return Math.abs(this.e) < EXP_ABS_CAP;
  }

  sign(): number {
    return this.m === 0 ? 0 : this.m < 0 ? -1 : 1;
  }

  abs(): BigNumber {
    return this.m < 0 ? new BigNumber(-this.m, this.e) : this;
  }

  neg(): BigNumber {
    return this.m === 0 ? this : new BigNumber(-this.m, this.e);
  }

  add(other: BigNumber): BigNumber {
    if (this.m === 0) return other;
    if (other.m === 0) return this;
    const de = this.e - other.e;
    if (de >= 16) return this;
    if (de <= -16) return other;
    if (de >= 0) {
      return new BigNumber(this.m + other.m * Math.pow(10, -de), this.e);
    }
    return new BigNumber(other.m + this.m * Math.pow(10, de), other.e);
  }

  sub(other: BigNumber): BigNumber {
    if (other.m === 0) return this;
    return this.add(other.neg());
  }

  mul(other: BigNumber): BigNumber {
    if (this.m === 0 || other.m === 0) return BigNumber.ZERO;
    return new BigNumber(this.m * other.m, this.e + other.e);
  }

  mulNum(n: number): BigNumber {
    if (n === 0 || this.m === 0) return BigNumber.ZERO;
    if (n === 1) return this;
    return new BigNumber(this.m * n, this.e);
  }

  div(other: BigNumber): BigNumber {
    if (other.m === 0) {
      if (this.m === 0) return BigNumber.ZERO;
      return new BigNumber(this.m < 0 ? -1 : 1, EXP_ABS_CAP);
    }
    if (this.m === 0) return BigNumber.ZERO;
    return new BigNumber(this.m / other.m, this.e - other.e);
  }

  divNum(n: number): BigNumber {
    if (n === 0) return this.m === 0 ? BigNumber.ZERO : new BigNumber(this.sign(), EXP_ABS_CAP);
    if (this.m === 0) return BigNumber.ZERO;
    return new BigNumber(this.m / n, this.e);
  }

  /** this^exp for real exp. Negative bases with non-integer exp → 0. */
  pow(exp: number): BigNumber {
    if (exp === 0) return BigNumber.ONE;
    if (this.m === 0) return exp > 0 ? BigNumber.ZERO : new BigNumber(1, EXP_ABS_CAP);
    if (this.m < 0 && !Number.isInteger(exp)) return BigNumber.ZERO;
    const log10 = Math.log(Math.abs(this.m)) * INV_LOG10 + this.e;
    const resultLog = log10 * exp;
    if (!Number.isFinite(resultLog)) {
      return new BigNumber(this.m < 0 && exp % 2 ? -1 : 1, EXP_ABS_CAP);
    }
    const e = Math.floor(resultLog);
    const m = Math.pow(10, resultLog - e);
    const sign = this.m < 0 && exp % 2 !== 0 ? -1 : 1;
    return new BigNumber(sign * m, e);
  }

  sqrt(): BigNumber {
    return this.m < 0 ? BigNumber.ZERO : this.pow(0.5);
  }

  log10(): number {
    if (this.m <= 0) return Number.NEGATIVE_INFINITY;
    return Math.log(this.m) * INV_LOG10 + this.e;
  }

  ln(): number {
    if (this.m <= 0) return Number.NEGATIVE_INFINITY;
    return Math.log(this.m) + this.e * LOG10;
  }

  floor(): BigNumber {
    if (this.m === 0) return this;
    if (this.e >= 15) return this;
    if (this.e < 0) return this.m >= 0 ? BigNumber.ZERO : BigNumber.from(-1);
    const v = this.toNumber();
    if (!Number.isFinite(v)) return this;
    return BigNumber.from(Math.floor(v));
  }

  cmp(other: BigNumber): number {
    if (this.m === 0 && other.m === 0) return 0;
    const sa = this.sign();
    const sb = other.sign();
    if (sa !== sb) return sa < sb ? -1 : 1;
    if (this.e !== other.e) {
      const expCmp = this.e < other.e ? -1 : 1;
      return sa < 0 ? -expCmp : expCmp;
    }
    if (this.m === other.m) return 0;
    return this.m < other.m ? -1 : 1;
  }

  eq(other: BigNumber): boolean {
    return this.cmp(other) === 0;
  }
  lt(other: BigNumber): boolean {
    return this.cmp(other) < 0;
  }
  lte(other: BigNumber): boolean {
    return this.cmp(other) <= 0;
  }
  gt(other: BigNumber): boolean {
    return this.cmp(other) > 0;
  }
  gte(other: BigNumber): boolean {
    return this.cmp(other) >= 0;
  }

  toNumber(): number {
    if (this.m === 0) return 0;
    if (this.e > 308) return this.m < 0 ? Number.NEGATIVE_INFINITY : Number.POSITIVE_INFINITY;
    if (this.e < -308) return 0;
    return this.m * Math.pow(10, this.e);
  }

  toJSON(): { m: number; e: number } {
    return { m: this.m, e: this.e };
  }

  toSci(digits = 3): string {
    if (this.m === 0) return "0";
    const d = Math.max(0, digits);
    return `${this.m.toFixed(d)}e${this.e}`;
  }

  toSuffix(digits = 2): string {
    return formatSuffix(this, digits);
  }

  toString(): string {
    return this.toSuffix();
  }
}

const SUFFIXES = [
  "",
  "K",
  "M",
  "B",
  "T",
  "Qa",
  "Qi",
  "Sx",
  "Sp",
  "Oc",
  "No",
  "Dc",
  "Ud",
  "Dd",
  "Td",
  "Qad",
  "Qid",
  "Sxd",
  "Spd",
  "Ocd",
  "Nod",
  "Vg",
  "UVg",
  "DVg",
  "TVg",
  "QaVg",
  "QiVg",
  "SxVg",
  "SpVg",
  "OcVg",
  "NoVg",
  "Tg",
];

export function formatSuffix(value: BigNumber, digits = 2): string {
  if (value.m === 0) return "0";
  const sign = value.m < 0 ? "-" : "";
  const abs = value.abs();
  if (abs.e < 3) {
    const n = abs.toNumber();
    if (n < 10) return sign + (Number.isInteger(n) ? String(n) : n.toFixed(Math.min(2, digits)));
    if (n < 100) return sign + (Number.isInteger(n) ? String(n) : n.toFixed(1));
    return sign + Math.floor(n).toLocaleString("en-US");
  }
  const group = Math.floor(abs.e / 3);
  if (group >= SUFFIXES.length) {
    return sign + abs.toSci(Math.max(2, digits));
  }
  const rem = abs.e - group * 3;
  const shown = abs.m * Math.pow(10, rem);
  const suffix = SUFFIXES[group];
  let body: string;
  if (shown >= 100) body = shown.toFixed(0);
  else if (shown >= 10) body = shown.toFixed(Math.min(1, digits));
  else body = shown.toFixed(Math.min(2, digits));
  body = body.replace(/\.0+$/, "").replace(/(\.\d*?)0+$/, "$1");
  return sign + body + suffix;
}

export function formatRate(value: BigNumber): string {
  return `${formatSuffix(value)}/s`;
}

export const bn = BigNumber.from;
